File: src/tachikoma/media.py
"""Media handling for Telegram channel.

Provides media descriptor table, download, description building,
file naming, and a bootstrap hook for temp folder management.
"""

import time
import uuid
from collections.abc import Callable
from pathlib import Path
from typing import Any, NamedTuple

from aiogram import Bot
from loguru import logger

from tachikoma.bootstrap import BootstrapContext

_log = logger.bind(component="media")

# Constants
MEDIA_TEMP_DIR = Path("/tmp/tachikoma-media")
TELEGRAM_MAX_FILE_SIZE = 20 * 1024 * 1024  # 20 MB
MEDIA_CLEANUP_DAYS = 30


class MediaTooLargeError(Exception):
    """Raised when a media file exceeds the Telegram bot download limit."""

    def __init__(self, file_size: int) -> None:
        self.file_size = file_size
        super().__init__(
            f"File too large to download ({file_size / (1024 * 1024):.1f} MB). "
            f"Telegram bots can only download files up to 20 MB."
        )


class MediaDescriptor(NamedTuple):
    """Descriptor for a supported media type."""

    field_name: str
    accessor: Callable[[Any], Any]
    resolve_extension: Callable[[Any], str]
    label: str
    build_metadata: Callable[[Any], list[str]]


# --- Metadata builders ---


def _photo_metadata(media: Any) -> list[str]:
    return [
        f"{media.width} × {media.height}",
        _format_file_size(media.file_size),
    ]


def _voice_metadata(media: Any) -> list[str]:
    lines = [f"{media.duration} seconds"]
    if media.mime_type:
        lines.append(media.mime_type)
    return lines


def _audio_metadata(media: Any) -> list[str]:
    parts: list[str] = []
    if media.title:
        parts.append(media.title)
    if media.performer:
        parts.append(media.performer)
    parts.append(f"{media.duration} seconds")
    return parts


def _document_metadata(media: Any) -> list[str]:
    lines: list[str] = []
    if media.file_name:
        lines.append(media.file_name)
    if media.mime_type:
        lines.append(media.mime_type)
    if media.file_size is not None:
        lines.append(_format_file_size(media.file_size))
    return lines


def _sticker_metadata(media: Any) -> list[str]:
    lines: list[str] = []
    if media.emoji:
        lines.append(f"Emoji: {media.emoji}")
    if media.is_animated:
        lines.append("Format: animated (.tgs)")
    elif media.is_video:
        lines.append("Format: video (.webm)")
    else:
        lines.append("Format: regular (.webp)")
    return lines


def _video_metadata(media: Any) -> list[str]:
    return [
        f"{media.duration} seconds",
        f"{media.width} × {media.height}",
    ]


def _video_note_metadata(media: Any) -> list[str]:
    return [
        f"{media.duration} seconds",
        f"Diameter: {media.length}px",
    ]


def _animation_metadata(media: Any) -> list[str]:
    return [
        f"{media.duration} seconds",
        f"{media.width} × {media.height}",
    ]


# --- Extension resolvers ---


def _extension_from_filename(media: Any, fallback: str) -> str:
    """Extract file extension from media's file_name, or return fallback."""
    file_name = getattr(media, "file_name", None)
    if file_name:
        ext = Path(file_name).suffix
        if ext:
            return ext
    return fallback


def _sticker_extension(media: Any) -> str:
    if media.is_animated:
        return ".tgs"
    if media.is_video:
        return ".webm"
    return ".webp"


# --- Descriptor table ---
# Order matters: more specific types before generic ones.
# animation before document (animations set both fields),
# video_note before video (defensive),
# document last (most generic).

MEDIA_DESCRIPTORS: list[MediaDescriptor] = [
    MediaDescriptor(
        field_name="animation",
        accessor=lambda msg: msg.animation,
        resolve_extension=lambda _: ".mp4",
        label="Animation",
        build_metadata=_animation_metadata,
    ),
    MediaDescriptor(
        field_name="sticker",
        accessor=lambda msg: msg.sticker,
        resolve_extension=_sticker_extension,
        label="Sticker",
        build_metadata=_sticker_metadata,
    ),
    MediaDescriptor(
        field_name="video_note",
        accessor=lambda msg: msg.video_note,
        resolve_extension=lambda _: ".mp4",
        label="Video note",
        build_metadata=_video_note_metadata,
    ),
    MediaDescriptor(
        field_name="photo",
        accessor=lambda msg: msg.photo[-1] if msg.photo else None,
        resolve_extension=lambda _: ".jpg",
        label="Photo",
        build_metadata=_photo_metadata,
    ),
    MediaDescriptor(
        field_name="voice",
        accessor=lambda msg: msg.voice,
        resolve_extension=lambda _: ".ogg",
        label="Voice message",
        build_metadata=_voice_metadata,
    ),
    MediaDescriptor(
        field_name="video",
        accessor=lambda msg: msg.video,
        resolve_extension=lambda m: _extension_from_filename(m, ".mp4"),
        label="Video",
        build_metadata=_video_metadata,
    ),
    MediaDescriptor(
        field_name="audio",
        accessor=lambda msg: msg.audio,
        resolve_extension=lambda m: _extension_from_filename(m, ".mp3"),
        label="Audio file",
        build_metadata=_audio_metadata,
    ),
    MediaDescriptor(
        field_name="document",
        accessor=lambda msg: msg.document,
        resolve_extension=lambda m: _extension_from_filename(m, ""),
        label="Document",
        build_metadata=_document_metadata,
    ),
]


def resolve_media(message: Any) -> tuple[Any, MediaDescriptor] | None:
    """Resolve the first matching media descriptor for a message.

    Iterates MEDIA_DESCRIPTORS in priority order and returns the
    first (media_object, descriptor) where the accessor finds a
    non-None value. Returns None if no media matches.
    """
    for descriptor in MEDIA_DESCRIPTORS:
        media_obj = descriptor.accessor(message)
        if media_obj is not None:
            return (media_obj, descriptor)
    return None


def generate_media_filename(descriptor: MediaDescriptor, media_object: Any) -> str:
    """Generate a unique filename for a downloaded media file.

    Types with an original file_name (documents, audio, video) preserve
    it appended to a UUID. Others use UUID + type-appropriate extension.
    """
    unique_id = uuid.uuid4().hex[:12]
    original_name = getattr(media_object, "file_name", None)

    if original_name:
        return f"{unique_id}-{original_name}"

    ext = descriptor.resolve_extension(media_object)
    return f"{unique_id}{ext}"


async def download_media(bot: Bot, media_object: Any, dest_path: Path) -> Path:
    """Download a media file from Telegram.

    Pre-checks file size against TELEGRAM_MAX_FILE_SIZE when available.
    Raises MediaTooLargeError if the file exceeds the limit.
    Raises TelegramAPIError on download failure.
    """
    file_size = getattr(media_object, "file_size", None)
    if file_size is not None and file_size > TELEGRAM_MAX_FILE_SIZE:
        raise MediaTooLargeError(file_size)

    await bot.download(media_object, destination=dest_path)
    return dest_path


def build_description(
    label: str,
    metadata_lines: list[str],
    file_path: Path,
    caption: str | None,
) -> str:
    """Build a natural-language description for a downloaded media file.

    Format:
        "The user sent a {label} ({metadata}).
         The file is saved at {file_path}"
         [Optional: 'The user said: "{caption}"']
    """
    metadata_str = ", ".join(metadata_lines)
    parts = [f"The user sent a {label} ({metadata_str})."]
    parts.append(f"The file is saved at {file_path}")

    if caption:
        parts.append(f'The user said: "{caption}"')

    return "\n".join(parts)


def _format_file_size(size: int | None) -> str:
    """Format file size in human-readable form."""
    if size is None:
        return "size unknown"
    if size < 1024:
        return f"{size} B"
    if size < 1024 * 1024:
        return f"{size / 1024:.0f} KB"
    return f"{size / (1024 * 1024):.1f} MB"


# --- Bootstrap hook ---


async def media_hook(ctx: BootstrapContext) -> None:
    """Bootstrap hook: ensure media temp directory exists and clean old files.

    Per DES-003, this hook is defined in the media module and registered
    in __main__.py. It is idempotent — safe to run every launch.
    """
    _log.debug("Ensuring media temp directory exists: path={path}", path=str(MEDIA_TEMP_DIR))

    MEDIA_TEMP_DIR.mkdir(parents=True, exist_ok=True)

    # Clean files older than MEDIA_CLEANUP_DAYS
    now = time.time()
    cutoff = now - (MEDIA_CLEANUP_DAYS * 24 * 60 * 60)
    cleaned = 0
    errors = 0

    for item in MEDIA_TEMP_DIR.iterdir():
        if not item.is_file():
            continue

        try:
            if item.stat().st_mtime < cutoff:
                item.unlink()
                cleaned += 1
        except OSError:
            _log.warning("Failed to delete old media file: path={path}", path=str(item))
            errors += 1

    _log.info(
        "Media temp directory ready: path={path}, cleaned={cleaned}, errors={errors}",
        path=str(MEDIA_TEMP_DIR),
        cleaned=cleaned,
        errors=errors,
    )

File: src/extensions/telegram/media.ts
import { randomBytes } from "node:crypto";
import { mkdir, readdir, stat, unlink, writeFile } from "node:fs/promises";
import { dirname, extname, join } from "node:path";
import type { Message } from "grammy/types";

import type { MediaAttachment, MediaKind } from "../../domain/message.ts";
import type { Logger } from "../../log.ts";

// Telegram bots can only download files up to 20 MB.
export const TELEGRAM_MAX_DOWNLOAD_BYTES = 20 * 1024 * 1024;

const MEDIA_RETENTION_DAYS = 30;

export type MediaMessage = Pick<
  Message,
  "animation" | "sticker" | "video_note" | "photo" | "voice" | "video" | "audio" | "document"
>;

export interface ResolvedMedia {
  kind: MediaKind;
  label: string;
  fileId: string;
  fileSize: number | null;
  fileName: string | null;
  extension: string;
  mimeType: string | null;
  metadata: Record<string, unknown>;
  /** Human-readable facts for the attachment description, e.g. "800 × 600". */
  summary: string[];
}

export class MediaTooLargeError extends Error {
  readonly fileSize: number;

  constructor(fileSize: number) {
    super(
      `File too large to download (${(fileSize / (1024 * 1024)).toFixed(1)} MB). ` +
        "Telegram bots can only download files up to 20 MB.",
    );
    this.fileSize = fileSize;
  }
}

/**
 * Resolve the first matching media payload on a message, in priority order:
 * animation before document (animations set both fields), video_note before
 * video (defensive), document last (most generic).
 */
export const resolveMedia = (message: MediaMessage): ResolvedMedia | null => {
  if (message.animation != null) {
    const media = message.animation;

    return {
      kind: "animation",
      label: "Animation",
      fileId: media.file_id,
      fileSize: media.file_size ?? null,
      fileName: media.file_name ?? null,
      extension: ".mp4",
      mimeType: media.mime_type ?? "video/mp4",
      metadata: { width: media.width, height: media.height, duration: media.duration },
      summary: [`${media.duration} seconds`, `${media.width} × ${media.height}`],
    };
  }

  if (message.sticker != null) {
    const media = message.sticker;
    const format = media.is_animated
      ? { extension: ".tgs", mimeType: "application/x-tgsticker", description: "animated (.tgs)" }
      : media.is_video
        ? { extension: ".webm", mimeType: "video/webm", description: "video (.webm)" }
        : { extension: ".webp", mimeType: "image/webp", description: "regular (.webp)" };

    return {
      kind: "sticker",
      label: "Sticker",
      fileId: media.file_id,
      fileSize: media.file_size ?? null,
      fileName: null,
      extension: format.extension,
      mimeType: format.mimeType,
      metadata: {
        ...(media.emoji != null ? { emoji: media.emoji } : {}),
        format: format.description,
      },
      summary: [
        ...(media.emoji != null ? [`Emoji: ${media.emoji}`] : []),
        `Format: ${format.description}`,
      ],
    };
  }

  if (message.video_note != null) {
    const media = message.video_note;

    // The domain has no dedicated video-note kind — closest is video.
    return {
      kind: "video",
      label: "Video note",
      fileId: media.file_id,
      fileSize: media.file_size ?? null,
      fileName: null,
      extension: ".mp4",
      mimeType: "video/mp4",
      metadata: { duration: media.duration, diameter: media.length },
      summary: [`${media.duration} seconds`, `Diameter: ${media.length}px`],
    };
  }

  const photo = message.photo?.at(-1);
  if (photo != null) {
    return {
      kind: "photo",
      label: "Photo",
      fileId: photo.file_id,
      fileSize: photo.file_size ?? null,
      fileName: null,
      extension: ".jpg",
      mimeType: "image/jpeg",
      metadata: { width: photo.width, height: photo.height },
      summary: [`${photo.width} × ${photo.height}`, formatFileSize(photo.file_size ?? null)],
    };
  }

  if (message.voice != null) {
    const media = message.voice;

    return {
      kind: "voice",
      label: "Voice message",
      fileId: media.file_id,
      fileSize: media.file_size ?? null,
      fileName: null,
      extension: ".ogg",
      mimeType: media.mime_type ?? "audio/ogg",
      metadata: { duration: media.duration },
      summary: [`${media.duration} seconds`, ...(media.mime_type != null ? [media.mime_type] : [])],
    };
  }

  if (message.video != null) {
    const media = message.video;

    return {
      kind: "video",
      label: "Video",
      fileId: media.file_id,
      fileSize: media.file_size ?? null,
      fileName: media.file_name ?? null,
      extension: extensionFromName(media.file_name, ".mp4"),
      mimeType: media.mime_type ?? null,
      metadata: { width: media.width, height: media.height, duration: media.duration },
      summary: [`${media.duration} seconds`, `${media.width} × ${media.height}`],
    };
  }

  if (message.audio != null) {
    const media = message.audio;

    return {
      kind: "audio",
      label: "Audio file",
      fileId: media.file_id,
      fileSize: media.file_size ?? null,
      fileName: media.file_name ?? null,
      extension: extensionFromName(media.file_name, ".mp3"),
      mimeType: media.mime_type ?? null,
      metadata: {
        duration: media.duration,
        ...(media.title != null ? { title: media.title } : {}),
        ...(media.performer != null ? { performer: media.performer } : {}),
      },
      summary: [
        ...(media.title != null ? [media.title] : []),
        ...(media.performer != null ? [media.performer] : []),
        `${media.duration} seconds`,
      ],
    };
  }

  if (message.document != null) {
    const media = message.document;

    return {
      kind: "document",
      label: "Document",
      fileId: media.file_id,
      fileSize: media.file_size ?? null,
      fileName: media.file_name ?? null,
      extension: extensionFromName(media.file_name, ""),
      mimeType: media.mime_type ?? null,
      metadata: { ...(media.file_name != null ? { fileName: media.file_name } : {}) },
      summary: [
        ...(media.file_name != null ? [media.file_name] : []),
        ...(media.mime_type != null ? [media.mime_type] : []),
        ...(media.file_size != null ? [formatFileSize(media.file_size)] : []),
      ],
    };
  }

  return null;
};

/** Unique filename: original names are preserved behind a random prefix. */
export const generateMediaFilename = (
  media: Pick<ResolvedMedia, "fileName" | "extension">,
): string => {
  const unique = randomBytes(6).toString("hex");

  return media.fileName != null ? `${unique}-${media.fileName}` : `${unique}${media.extension}`;
};

export const buildAttachment = (media: ResolvedMedia, path: string): MediaAttachment => ({
  kind: media.kind,
  path,
  ...(media.mimeType != null ? { mimeType: media.mimeType } : {}),
  description: `${media.label} (${media.summary.join(", ")})`,
  metadata: { ...media.metadata, ...(media.fileSize != null ? { fileSize: media.fileSize } : {}) },
});

export interface FileApi {
  getFile(fileId: string): Promise<{ file_path?: string }>;
}

/** Download a Telegram file to disk, pre-checking the bot download size limit. */
export const downloadMedia = async (
  api: FileApi,
  token: string,
  media: Pick<ResolvedMedia, "fileId" | "fileSize">,
  destPath: string,
): Promise<void> => {
  if (media.fileSize != null && media.fileSize > TELEGRAM_MAX_DOWNLOAD_BYTES) {
    throw new MediaTooLargeError(media.fileSize);
  }

  const file = await api.getFile(media.fileId);
  if (file.file_path == null) throw new Error("Telegram returned no file path for download");

  const response = await fetch(`https://api.telegram.org/file/bot${token}/${file.file_path}`);
  if (!response.ok) throw new Error(`File download failed with status ${response.status}`);

  await mkdir(dirname(destPath), { recursive: true });
  await writeFile(destPath, Buffer.from(await response.arrayBuffer()));
};

/** Bootstrap: ensure the media directory exists and prune stale downloads. */
export const ensureMediaDir = async (mediaDir: string, log: Logger): Promise<void> => {
  await mkdir(mediaDir, { recursive: true });

  const cutoff = Date.now() - MEDIA_RETENTION_DAYS * 24 * 60 * 60 * 1000;
  let cleaned = 0;

  for (const entry of await readdir(mediaDir, { withFileTypes: true })) {
    if (!entry.isFile()) continue;

    const path = join(mediaDir, entry.name);

    try {
      if ((await stat(path)).mtimeMs < cutoff) {
        await unlink(path);
        cleaned += 1;
      }
    } catch (error) {
      log.warn({ err: error, path }, "failed to clean old media file");
    }
  }

  log.debug({ mediaDir, cleaned }, "media directory ready");
};

const extensionFromName = (fileName: string | undefined, fallback: string): string => {
  if (fileName != null) {
    const extension = extname(fileName);
    if (extension !== "") return extension;
  }

  return fallback;
};

const formatFileSize = (size: number | null): string => {
  if (size == null) return "size unknown";
  if (size < 1024) return `${size} B`;
  if (size < 1024 * 1024) return `${Math.round(size / 1024)} KB`;

  return `${(size / (1024 * 1024)).toFixed(1)} MB`;
};
